src/plugins/browse/mod.rs

```diff
@@ -59,66 +72,98 @@ impl Error for BrowseNoArgError {}
 fn chunk_text(text: &str, chunk_size: usize) -> Vec<String> {
+    let bpe = cl100k_base().unwrap();
+    let tokens = bpe.encode_ordinary(text);
     let mut chunks = vec![];
-    let mut current_chunk = String::new();
+    let mut token_vec: Vec<_> = tokens.to_vec();
 
-    for word in wrap(text, chunk_size) {
-        if current_chunk.len() + word.len() > chunk_size {
-            chunks.push(current_chunk.trim().to_owned());
-            current_chunk = String::new();
-        }
-        current_chunk.push_str(&word);
-    }
-    if !current_chunk.is_empty() {
-        chunks.push(current_chunk.trim().to_owned());
+    while !token_vec.is_empty() {
+        let drain_to = std::cmp::min(token_vec.len(), chunk_size);
+        let token_chunk = token_vec.drain(0..drain_to).collect::<Vec<_>>();
+
+        let chunk = bpe.decode(token_chunk).unwrap();
+        chunks.push(chunk);
     }
+
     chunks
 }
 
-pub async fn browse_url(ctx: &mut CommandContext, args: Vec<ScriptValue>) -> Result<ScriptValue, Box<dyn Error>> {
+pub async fn browse_url(
+    ctx: &mut CommandContext,
+    args: Vec<ScriptValue>,
+) -> Result<ScriptValue, Box<dyn Error>> {
     let browse_info = ctx.plugin_data.get_data("Browse")?;
 
     let params: [(&str, &str); 0] = [];
-    let url: String = args.get(0).ok_or(BrowseNoArgError)?.clone().try_into()?;   
+    let url: String = args.get(0).ok_or(BrowseNoArgError)?.clone().try_into()?;
 
-    let body = invoke::<String>(browse_info, "browse", BrowseRequest {
-        url: url.to_string(),
-        params: params.iter()
-            .map(|el| (el.0.to_string(), el.1.to_string()))
-            .collect::<Vec<_>>()
-    }).await?;
+    // let body = invoke::<String>(
+    //     browse_info,
+    //     "browse",
+    //     BrowseRequest {
+    //         url: url.to_string(),
+    //         params: params
+    //             .iter()
+    //             .map(|el| (el.0.to_string(), el.1.to_string()))
+    //             .collect::<Vec<_>>(),
+    //     },
+    // )
+    // .await?;
 
-    let content = extract_text_from_html(&body);
+    let content = headless::get_content_headless(&url).await?;
+
+    // let content = extract_text_from_html(&body);
 
     let mut summarized_content = String::new();
-    let chunks = chunk_text(&content, 11000);
+    let chunks = chunk_text(&content, 3000);
 
     let chunk_count = chunks.len();
     let summary_prompt = match chunk_count {
         0..=2 => "Create a paragraph summary.",
-        _ => "Create a two-sentence summary."
-    }.to_string();
+        _ => "Create a two-sentence summary.",
+    }
+    .to_string();
 
     for (ind, chunk) in chunks.iter().enumerate() {
-        println!("{} {} / {}", "Summarizing Chunk".green(), ind + 1, chunks.len());
+        println!(
+            "{} {} / {}",
+            "Summarizing Chunk".green(),
+            ind + 1,
+            chunks.len()
+        );
 
```

Comment:
summarize all texts on a page may not be necessary for every run
I tried to check the coherence of the text on the first "virtual pdf" page with a local Bert model, if Bert finds it good enough, we don't bother processing the remainder of the page, if not, we pay for external api to summarize the whole page
it's too much for me to swallow at this time